src/hotspot/share/gc/shared/barrierSetNMethod.cpp

```diff
@@ -38,11 +40,18 @@
 #include "runtime/threads.hpp"
 #include "utilities/debug.hpp"
 
-class LoadPhantomOopClosure : public OopClosure {
+class OopKeepaliveClosure : public OopClosure {
+  CollectedHeap* _heap;
+
 public:
+  OopKeepaliveClosure()
+    : _heap(Universe::heap()) {}
+
   virtual void do_oop(oop* p) {
-    NativeAccess<ON_PHANTOM_OOP_REF>::oop_load(p);
+    oop obj = NativeAccess<ON_PHANTOM_OOP_REF | AS_NO_KEEPALIVE>::oop_load(p);
+    _heap->keep_alive(obj);
   }
+
   virtual void do_oop(narrowOop* p) { ShouldNotReachHere(); }
 };
```

Comment:
* Keepalive should probably be KeepAlive. (I realize that AS_NO_KEEPALIVE doesn't split the two words, but that one is the outlier in the HotSpot code base)
* Seems unnecessary to cache the heap pointer
* Could this class be moved into the BarrierSetNMethod scope. Or, alternatively give it a more descriptive name stating that this is for nmethods and/or native phantom loads?
 I'm thinking about future name collisions and ODR.
* Could you add a comment stating why you do this dance with AS_NO_KEEPALIVE + keep_alive(oop)